src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs

```diff
@@ -182,6 +185,67 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
             }
         }
 
+        public string InstallWorkloadSet(string path)
+        {
+            return ModifyWorkloadSet(path, InstallAction.Install);
+        }
+
+        public void RollBackWorkloadSetInstallation()
+        {
+            if (_newWorkloadSetPath is not null)
+            {
+                ModifyWorkloadSet(_newWorkloadSetPath, InstallAction.Uninstall);
+                if (_previousWorkloadSetVersion is not null)
+                {
+                    File.WriteAllText(Path.Combine(_newWorkloadSetPath, "version.txt"), _previousWorkloadSetVersion);
+                    ModifyWorkloadSet(_newWorkloadSetPath, InstallAction.Repair);
+                }
+            }
+        }
+
+        private string ModifyWorkloadSet(string path, InstallAction installAction)
+        {
+            ReportPendingReboot();
+
+            // Rolling back a manifest update after a successful install is essentially a downgrade, which is blocked so we have to
+            // treat it as a special case and is different from the install failing and rolling that back, though depending where the install
+            // failed, it may have removed the old product already.
+            // Resolve the package ID for the manifest payload package
+            var featureBand = Path.GetFileName(Path.GetDirectoryName(path));
+            var version = File.ReadAllText(Path.Combine(path, "version.txt"));
+            string msiPackageId = GetManifestPackageId(new ManifestId("Microsoft.NET.Workloads"), new SdkFeatureBand(featureBand)).ToString();
+            string msiPackageVersion = version;
+
+            Log?.LogMessage($"Resolving Microsoft.NET.Workloads ({version}) to {msiPackageId} ({msiPackageVersion}).");
+
+            // Retrieve the payload from the MSI package cache.
+            MsiPayload msi = GetCachedMsiPayload(msiPackageId, msiPackageVersion, null);
+            VerifyPackage(msi);
+            DetectState state = DetectPackage(msi.ProductCode, out Version installedVersion);
+
+            if (installAction == InstallAction.Install)
+            {
+                // If we are installing the workload set, record its version and the previous version so that we can roll back if necessary.
+                _newWorkloadSetPath = path;
+                var parentDirectory = new DirectoryInfo(Path.GetDirectoryName(Path.GetDirectoryName(msi.MsiPath)));
+                _previousWorkloadSetVersion = parentDirectory.EnumerateDirectories().Select(d => d.Name).Where(n => !n.Equals(version))?.Max();
+            }
+
+            InstallAction plannedAction = PlanPackage(msi, state, installAction, installedVersion, out IEnumerable<string> relatedProducts);
+
+            if (plannedAction != InstallAction.None)
```

Comment:
If there is no planned action, you don't need to execute the package. You also have to elevate before you can updat the dependent. I'd move the execute package and update call into the IF block

Reply:
ExecutePackage basically just exits if plannedAction is InstallAction.None, and UpdateDependent Elevates separately, so it doesn't actually affect anything in this case, but I have no problem doing that.